crates/shared/src/token_list.rs

```diff
@@ -18,31 +51,46 @@ pub struct Token {
 }
 
 impl TokenList {
-    pub async fn from_url(url: impl IntoUrl, chain_id: u64, client: Client) -> Result<Self> {
-        let model: TokenListModel = client.get(url).send().await?.json().await?;
-        Ok(Self::from_tokens(model.tokens, chain_id))
-    }
+    pub async fn from_configuration(configuration: TokenListConfiguration) -> Self {
+        let tokens = Arc::new(RwLock::new(
+            configuration.tokens().await.unwrap_or_default(),
+        ));
 
-    fn from_tokens(tokens: Vec<TokenModel>, chain_id: u64) -> Self {
-        Self {
-            tokens: tokens
-                .into_iter()
-                .filter(|token| token.chain_id == chain_id)
-                .map(|token| (token.token.address, token.token))
-                .collect(),
+        // spawn a background task to regularly update token list
+        {
+            let tokens = tokens.clone();
+            let updater = async move {
+                loop {
+                    tokio::time::sleep(configuration.update_interval).await;
+
+                    if let Ok(new_tokens) = configuration.tokens().await {
+                        let mut w = tokens.write().unwrap();
+                        *w = new_tokens;
+                    }
```

Comment:
This is going to silently fail. Should be warning or error log.

Reply:
Agree. Even better, can we add a metric that counts success and failures of fetching the token list? Then we can setup Grafana alerts if we fail too many times in a row to update the token list!

Reply:
Added log in case of failure. I feel like having a metric with counters and alert for this is a bit overkill since I don't expect any failures here. Also I think we can add it later if it turns out I'm wrong about it.